### PR TITLE
Fixes for large datasets

### PR DESCRIPTION
**Pull request by mcsqr**
Some minor but somewhat helpful changes:
* This avoids invoking `S.values` which creates the dense matrix from S when `sparse_s = True` in the `aggregate` function. This avoids OOM errors for very large datasets (I've used it [the `aggregate` function] successfully on a 400k time-series dataset after the change).
* It also removes a warning coming from using the deprecated `sparse` argument. Backwards compatibility with old sklearn libraries is maintained.

**Excluded from this PR:**
* It gives a warning whenever Y_df has less time series than S_df at the end of the `aggregate` function, as current behaviour is quite unintuitive and leads to users' confusion.
[EDIT: Aaah, ok, it seems that this behaviour was introduced as a result of this PR here: Nixtla/hierarchicalforecast/pull/190, see also this issue: Nixtla/hierarchicalforecast/issues/211 . What was the reason again to move this line `Y_bottom_df = Y_bottom_df.groupby(['unique_id', 'ds'])['y'].sum().reset_index()` two lines up?]
[EDIT 2: Perhaps there should be an additional argument with an option whether to prepend the shorter time series with zeros, or not?]

### CODE REVIEW

**Comment by jmoralez** [comment]
Hey. Sorry for taking so long to review, I'll take a look now.

The errors weren't exactly related to python 3.7 but rather scikit-learn >= 1.2, [which was released in Dec 2022](https://pypi.org/project/scikit-learn/1.2.0/), not so long ago to stop supporting it. Can you please use something like the following instead?
```python
try:
    encoder = OneHotEncoder(categories=categories, sparse_output=sparse_s, dtype=np.float32)
except TypeError:  # sklearn < 1.2
    encoder = OneHotEncoder(categories=categories, sparse=sparse_s, dtype=np.float32)
```

**Comment by mcsqr** [comment]
Hi Jose, yeah, thanks for the tip. I did it.

Note: the currently biggest point of discussion is the last bullet point in the PR description. It seems that the change you (Fede) have introduced in PR #190 is very unintuitive to users. I'm not sure what's the right way to handle this, but you have some issues open and I also had complaints from my collaborators. The warning I added is just a small plaster on a big wound. But this could also be taken out of this PR if desired.

File: hierarchicalforecast/utils.py [hunk 1]
```diff
@@ -70,22 +70,29 @@ def cov2corr(cov, return_std=False):
...
-    encoder = OneHotEncoder(categories=categories, sparse=False, dtype=np.float32)
+
+    try:
+        encoder = OneHotEncoder(categories=categories, sparse_output=sparse_s, dtype=np.float32)
+    except TypeError:  # sklearn < 1.2
+        encoder = OneHotEncoder(categories=categories, sparse=sparse_s, dtype=np.float32)
+
     S = encoder.fit_transform(S_df).T
+    # TODO: This likely still instantiates a dense matrix:
     S = pd.DataFrame(S, index=chain(*categories), columns=cats_bottom)
```
Comment (jmoralez) [inline hierarchicalforecast/utils.py:87]: I think we should use the [recommended constructor](https://pandas.pydata.org/docs/user_guide/sparse.html#interaction-with-scipy-sparse), something like:
```python
if sparse_s:
    df_constructor = pd.DataFrame.sparse.from_spmatrix
else:
    df_constructor = pd.DataFrame
S = df_constructor(S, index=chain(*categories), columns=cats_bottom)
```

File: hierarchicalforecast/utils.py [hunk 2]
```diff
@@ -271,7 +285,16 @@ def aggregate(
...
                 y = np.concatenate([y_agg, y_bottom], axis=0)))
-    Y_df = Y_df.set_index('unique_id').dropna()
+
+    Y_df = Y_df.set_index("unique_id").dropna()
+
+    if S_df.index.nunique() != Y_df.index.nunique():
+        print(
+            "Warning: Some time series were dropped from Y_df. "
+            "Are you sure your time series don't have missing entries?"
+        )
```
Comment (jmoralez) [inline hierarchicalforecast/utils.py:295]: I think we can leave this for a follow up PR where we try to directly address #211

File: hierarchicalforecast/utils.py [hunk 3]
```diff
@@ -255,9 +266,12 @@ def aggregate(
 
     #------------------------------- Aggregation -------------------------------#
     n_agg = S_df.shape[0] - S_df.shape[1]
-    Agg = S_df.values[:n_agg, :]
-    y_bottom = balanced_df.y.values
+    if sparse_s:
+        Agg = S_df.sparse.to_coo().tocsr()[:n_agg, :]
```
Comment (jmoralez) [inline hierarchicalforecast/utils.py:270]: Can you add a test in the notebook to verify that this is working as expected? Could be just a small example with sparse_s set to both True and False and verify that the outputs are the same.

**Review by jmoralez — CHANGES_REQUESTED** [review]
Just some minor comments. Also can you please revert the changes to python 3.7? The library is still compatible with it and we want to support it for a little big longer.

File: hierarchicalforecast/utils.py (hunk 2 above)
Comment (mcsqr) [reply hierarchicalforecast/utils.py:295]: Yeah, I can remove it from this PR. Note however that it's a major annoyance for the users.

File: hierarchicalforecast/utils.py (hunk 1 above)
Comment (mcsqr) [reply hierarchicalforecast/utils.py]: Done

File: hierarchicalforecast/utils.py (hunk 2 above)
Comment (mcsqr) [reply hierarchicalforecast/utils.py]: Done

File: hierarchicalforecast/utils.py (hunk 3 above)
Comment (mcsqr) [reply hierarchicalforecast/utils.py:275]: Done

**Comment by mcsqr** [comment]
@jmoralez I think all done.

**Review by jmoralez — APPROVED** [review]
Thanks a lot! LGTM

**Pull request merged**